File: answers/lintcode/Sort Colors.cpp
class Solution{
public:
    /**
     * @param nums: A list of integer which is 0, 1 or 2 
     * @return: nothing
     */    
    void sortColors(vector<int> &nums) {
        // write your code here
		size_t zero = 0;
		size_t two = nums.size() - 1;
		for (size_t i = 0; i <= two && i < nums.size(); ++ i) {
			if (0 == nums[i]) {
				nums[i] = nums[zero];
				nums[zero] = 0;
				++ zero;
			}
			else if (2 == nums[i]) {
				nums[i] = nums[two];
				nums[two] = 2;
				-- two;
				// move iterator backward, since the new item has not been examined
				-- i;
			}
		}
    }
};

File: answers/leetcode/Sort Colors/Sort Colors.java
class Solution {
    public void sortColors(int[] nums) {
        int left = 0;
        int right = nums.length - 1;
        for (int i = 0; i < nums.length && i <= right; i++) {
            int temp = nums[i];
            if (nums[i] == 0 && i > left) {
                nums[i] = nums[left];
                nums[left] = temp;
                left++;
                i--;
            } else if (nums[i] == 2) {
                nums[i] = nums[right];
                nums[right] = temp;
                right--;
                i--;
            }
        }
    }
}